llbcache/resolveimage.go

```diff
@@ -0,0 +1,98 @@
+package llbcache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/opencontainers/go-digest"
+)
+
+// ResolveImageFunc is used to populate the cache for an image.
+type ResolveImageFunc func() (digest.Digest, []byte, error)
+
+// ResolveImageCache provides an interface to lookup resolved images in a cache
+// which allows multiple (and/or parallel) lookups for the same image to be
+// dedup'd and cached.
+type ResolveImageCache interface {
+	// Lookup will look in the cache to find a previously attempted resolve and
+	// will return the previous resolution.  If the resolve is a duplicate and
+	// in-flight then Lookup will wait for the original request to finish then
+	// return the result.  If the the lookup was not in the cache we will
+	// call the provided resolver to complete the lookup and cache the results.
+	Lookup(ctx context.Context, ref string, opt llb.ResolveImageConfigOpt, resolver ResolveImageFunc) (digest.Digest, []byte, error)
+}
+
+type resolveImageCacheKey struct {
+	resolveType llb.ResolverType
+	ref         string
+	os          string
+	arch        string
+	variant     string
+	mode        string
+	store       llb.ResolveImageConfigOptStore
+}
+
+type resolveImageCacheValue struct {
+	digest   digest.Digest
+	config   []byte
+	err      error
+	inflight chan struct{}
+}
+
+// NewResolveImageCache will return a new cache for resolved images.
+func NewResolveImageCache() ResolveImageCache {
+	return &resolveImageCache{
+		cache: map[resolveImageCacheKey]*resolveImageCacheValue{},
+	}
+}
+
+type resolveImageCache struct {
+	mu    sync.Mutex
+	cache map[resolveImageCacheKey]*resolveImageCacheValue
+}
+
+func (r *resolveImageCache) Lookup(ctx context.Context, ref string, opt llb.ResolveImageConfigOpt, resolver ResolveImageFunc) (digest.Digest, []byte, error) {
```

Comment:
This should be easy to unit test but it's so simple that I'm not sure the payoff is very high. Your call.

Reply:
Yeah, this is exercised already via other unit tests, not sure we will get much value out of testing it directly.